### PR TITLE
02-client routing: implement `VerifyUpgradeAndUpdateState` in light client modules

### PR DESCRIPTION
**Pull request by crodriguezvega**
## Description

closes: #XXXX

### Commit Message / Changelog Entry

```text
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/keeper/msg_server.go
Comment (crodriguezvega) [inline modules/core/keeper/msg_server.go:75]: @colin-axner @damiannolan I think just passing doing `.Value` here and passing the bytes of the proton-encoded client and consensus states might not be enough. See test failing [here](https://github.com/cosmos/ibc-go/actions/runs/7849754768/job/21423773040?pr=5827), it looks like if we just pass the bytes we miss the type URL information that we need to unmarshal the bytes.

File: modules/light-clients/07-tendermint/light_client_module.go
Comment (crodriguezvega) [inline modules/light-clients/07-tendermint/light_client_module.go:319]: @colin-axner @damiannolan This validation fails if the trust level is not within [0, 1/3] (see test failing [here](https://github.com/cosmos/ibc-go/actions/runs/7849754768/job/21423772705?pr=5827)). However, the upgraded client state that the counterparty chain commits should have all the custom fields zeroed out [since that's what we verify in `VerifyUpgradeAndUpdateState` function in the `ClientState`](https://github.com/cosmos/ibc-go/blob/d957ef80d72a260af3f6e66bbf5d7bea85473e2b/modules/light-clients/07-tendermint/upgrade.go#L73). So I think we either cannot do this validation here or we need to validate only the fields that shouldn't be zeroed out.

File: modules/core/keeper/msg_server.go
Comment (damiannolan) [reply modules/core/keeper/msg_server.go:75]: See https://github.com/cosmos/ibc-go/pull/5827/commits/3e311639ef002c2272ed02e11cbd5aee1bc543d3

We should unmarshal to concrete types and not interfaces within the light client module.

File: modules/light-clients/07-tendermint/light_client_module.go
Comment (damiannolan) [reply modules/light-clients/07-tendermint/light_client_module.go]: I saw this failing test when working on the one above. I didn't dig any deeper on this one yet. But from what you're saying here it sounds correct. I'm not sure exactly what pieces of code we'll need to change yet, can follow up o this one next week! :)

File: modules/core/02-client/keeper/client.go
Comment (damiannolan) [inline modules/core/02-client/keeper/client.go:127]: Isn't this an API breaking change? shouldn't the API remain the same here in the 02-client keeper and continue to pass the `exported` interfaces.

We can then pass `.Value` bytes when calling into the light client module.

File: modules/core/keeper/msg_server.go
Comment (damiannolan) [inline modules/core/keeper/msg_server.go:75]: essentially push this down a level into 02-client

File: modules/core/keeper/msg_server_test.go
Comment (damiannolan) [inline modules/core/keeper/msg_server_test.go:891]: decided to change this while I was here, makes no difference but kinda looks weird how it was 🧹 

I think these tests could be improved or cleaned up some more. But maybe outside the scope of this PR

File: modules/core/02-client/keeper/client.go
Comment (crodriguezvega) [reply modules/core/02-client/keeper/client.go:127]: It is API breaking, but I think that this refactoring will be API breaking anyway and will be released in a major release. I basically did here [a similar change as was done in `CreateClient`](https://github.com/cosmos/ibc-go/blob/f4ee20bf8c9139c7806c6410785ae37073bc7702/modules/core/02-client/keeper/client.go#L20).

File: modules/core/keeper/msg_server.go
Comment (crodriguezvega) [reply modules/core/keeper/msg_server.go:75]: Thanks, Damian. I changed this to unmarshal to interface because [the tests are generally using `client types.MarshalClientState`](https://github.com/cosmos/ibc-go/blob/055a3e39c3da309e013ec4763f78d33b74be59c9/modules/core/02-client/keeper/client_test.go#L481), which [calls `cdc.MarshalInterface`](https://github.com/cosmos/ibc-go/blob/055a3e39c3da309e013ec4763f78d33b74be59c9/modules/core/02-client/types/encoding.go#L35), so when trying to unmarshal to the concrete type I would get [errors like the ones you see now in the PR](https://github.com/cosmos/ibc-go/actions/runs/7952186466/job/21706474023?pr=5827).

So I guess my question would be: do we expect the client state (and consensus state) received in the messages be marshalled as an interface or a concrete type? I am not sure if this question really makes sense, though. Maybe relayers will just send the bytes for the marshalled concrete type (it seems like [that's what hermes does](https://github.com/informalsystems/hermes/blob/f986dc82cff5aafef11befee44b1d2eb8d179ad9/crates/relayer/src/client_state.rs#L149-L152), although the [go relayer seems to marshal the interface type](https://github.com/cosmos/relayer/blob/11b7de318b5527b0f097475df1865fe472f76ff1/relayer/chains/cosmos/tx.go#L708-L716)?), so maybe we just need to change our tests to also marshal to the concrete type? Sorry, my protobuf foo is pretty weak, so I might just not making any sense, hehe. But having a bit of clarity on this would be helpful for me to unblock this work.

File: modules/light-clients/07-tendermint/light_client_module.go
Comment (damiannolan) [reply modules/light-clients/07-tendermint/light_client_module.go]: I removed the basic validation of client state and consensus state at the light client module level. ClientState is reconstructed within the method `VerifyUpgradeAndUpdateState` and validated there. 

There is still some test failures wrt encoding from what I can see, but I think they should be manageable. Is the plan to restructure tests to invoke the client module in this PR or to leave it for a follow up? 

Happy to help out where needed :)

File: modules/light-clients/08-wasm/light_client_module.go
Comment (colin-axner) [inline modules/light-clients/08-wasm/light_client_module.go:286]: godoc should probably be rewritten to be wasm specific, not a copy/paste of the API godoc

File: modules/light-clients/08-wasm/light_client_module.go
Comment (colin-axner) [inline modules/light-clients/08-wasm/light_client_module.go:310]: I wonder if we could unmarshal directly into the concrete type and drop the cast assertions that follow?

File: modules/light-clients/07-tendermint/light_client_module_test.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/light_client_module_test.go:265]: super nit: one thing I dislike about assigning the context to a variable rather than dynamically requesting it is that this variable has the incorrect block height set in the context. The `tc.malleate()` call will commit a block on chainA on the "success" case, resulting in the context having an out of date block header when calling `VerifyUpgradeAndUpdateState`. I guess it is fine in this test, but can lead to confusing errors if it needs to be accurate

Comment (crodriguezvega) [reply modules/light-clients/07-tendermint/light_client_module_test.go]: Got it. I changed it now.

File: modules/light-clients/08-wasm/light_client_module.go
Comment (crodriguezvega) [reply modules/light-clients/08-wasm/light_client_module.go]: Fixed.

File: modules/core/02-client/keeper/client_test.go
Comment (crodriguezvega) [inline modules/core/02-client/keeper/client_test.go:360]: I deleted this test because this is tested in the light client module.

File: modules/core/02-client/keeper/client_test.go
Comment (crodriguezvega) [inline modules/core/02-client/keeper/client_test.go:449]: I moved this test to the light client module, but can move it back here in the PR where `LatestHeight` is added to the light client module and then we can move the check back to 02-client.

File: modules/light-clients/08-wasm/light_client_module.go
Comment (crodriguezvega) [reply modules/light-clients/08-wasm/light_client_module.go:263]: As discussed on Monday, we need to unmarshal into the interface.

**Comment by crodriguezvega** [comment]
I think this is again ready for review after the discussion on Monday. There are a couple of tests that fail because we need to add the right value for the latest height in the create and upgrade client events, but that should be fixed in https://github.com/cosmos/ibc-go/pull/5866.

File: modules/core/keeper/msg_server.go
Comment (colin-axner) [inline modules/core/keeper/msg_server.go:72]:
```suggestion
        // Backwards Compatibility: the light client module is expecting
        // to unmarshal an interface so we marshal the client state Any
	upgradedClientState, err := k.cdc.Marshal(msg.ClientState)
```

File: modules/core/keeper/msg_server.go
Comment (colin-axner) [inline modules/core/keeper/msg_server.go:77]:
```suggestion
        // Backwards Compatibility: the light client module is expecting
        // to unmarshal an interface so we marshal the consensus state Any
	upgradedConsensusState, err := k.cdc.Marshal(msg.ConsensusState)
```

some sort of in line code doc like this can be useful

File: modules/light-clients/07-tendermint/light_client_module.go
Comment (colin-axner) [inline modules/light-clients/07-tendermint/light_client_module.go:301]: Add back to 02-client level once latest height pr is merged?

**Review by colin-axner — APPROVED** [review]
Thanks @crodriguezvega!!

**Comment by damiannolan** [comment]
Resolving conflicts.. 🏃🏻

File: modules/core/02-client/keeper/client.go
Comment (damiannolan) [reply modules/core/02-client/keeper/client.go:142]: discussed async, this check must be moved to the light client module, and this functionality should likely only exist in the tendermint client module.

At first glance, we do not have the `clientID` of the `upgradedClientState`. However, it should be the same `clientID` as the `clientState` (target). However, we need the latest height from the `upgradedClientState` concrete implementation here - this is the client which the counterparty stores in it upgrade path and we prove against within the tm client module.

File: modules/core/02-client/keeper/client_test.go
Comment (damiannolan) [inline modules/core/02-client/keeper/client_test.go:456]: this test needs to move given that we have removed the height check

File: modules/core/02-client/keeper/client_test.go
Comment (damiannolan) [reply modules/core/02-client/keeper/client_test.go:457]: https://github.com/cosmos/ibc-go/pull/5827#discussion_r1520067044

I guess it can just be removed entirely now?

File: modules/core/keeper/msg_server.go
Comment (AdityaSripal) [inline modules/core/keeper/msg_server.go:75]: It's a bit unfortunate that we're marshalling here and then unmarshalling later in the callstack. Seems wasteful...

**Review by AdityaSripal — COMMENTED** [review]
Looks good! I suppose we have to keep the interfaces for users the same way, but would be nice to eventually just pass everything through and unmarshal at the light client module level.

File: modules/core/02-client/keeper/client_test.go
Comment (crodriguezvega) [reply modules/core/02-client/keeper/client_test.go:457]: Sounds good. At the moment this check is performed in 07-tendermint and 08-wasm light client modules. Based on [your comment](https://github.com/cosmos/ibc-go/pull/5827/files#r1520066359), is that ok? Any reason why the check should not be done in 08-wasm?

File: modules/core/keeper/msg_server.go
Comment (damiannolan) [reply modules/core/keeper/msg_server.go:75]: I don't think this is necessary. I think this is symptom of how the tests were/are written and they just need to be refactored. After looking closer at the code and how it works, I think it should be possible to pass the `Any.Value` bytes along here and into the light client module and have it work.

Opened a draft which I'll pick up in the morning, but see here for happy path test working: https://github.com/cosmos/ibc-go/pull/5967/files#diff-cef346df04f3fb2455741328504cc2cbb06c7133cce6ede300347139fbdda74eR490

tldr; test code needs to define `*codectypes.Any` and pass their value bytes as args. We can then make the changes in the light client module to unmarshal to concrete types. When we reconstruct the clientState which is proven it has always used `MarshalInterface` given the concrete type and should continue to work like that. I think the tests just need more attention.

File: modules/core/02-client/keeper/client_test.go
Comment (charleenfei) [reply modules/core/02-client/keeper/client_test.go:457]: we can have these tests in the lcm modules for 07-tendermint and 08-wasm already, i think! is that fine @crodriguezvega?

Comment (crodriguezvega) [reply modules/core/02-client/keeper/client_test.go:457]: Yes, correct! The check and the tests are currently in the light client module. Happy to leave it there if that's the right place. :)

File: modules/core/02-client/keeper/client_test.go
Comment (damiannolan) [inline modules/core/02-client/keeper/client_test.go:365]:
```suggestion
				// upgrade height is at next block
```

File: modules/core/02-client/keeper/client_test.go
Comment (damiannolan) [inline modules/core/02-client/keeper/client_test.go:397]:
```suggestion
				// upgrade height is at next block
```

File: modules/core/02-client/keeper/client_test.go
Comment (damiannolan) [inline modules/core/02-client/keeper/client_test.go:430]:
```suggestion
				// upgrade height is at next block
```

File: modules/light-clients/08-wasm/light_client_module.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/light_client_module.go:295]: This code needs to be aligned with 07-tendermint impl, shame that this is not caught with tests.. hmmm

File: modules/light-clients/08-wasm/light_client_module_test.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/light_client_module_test.go:1]: And tests

**Review by damiannolan — CHANGES_REQUESTED** [review]
I missed 08-wasm changes needed to aligned with 07-tendermint from my encoding PR. 
I'll handle them today and then LGTM! 🙏🏻

File: modules/core/02-client/keeper/client_test.go
Comment (damiannolan) [reply modules/core/02-client/keeper/client_test.go:388]: can follow up this post merge, maybe? I think all these tests should be using `expError` and not `expPass` (bool)

**Review by damiannolan — APPROVED** [review]
LGTM!

**Pull request merged**
